cpg_workflows/dataproc_scripts/mt_to_es_free_of_dataproc.py

```diff
@@ -0,0 +1,304 @@
+    parser.add_argument('--index', help='ES index name', required=True)
+    parser.add_argument('--flag', help='ES index "DONE" file path')
+    args = parser.parse_args()
+    logging.basicConfig(level=logging.INFO)
+
+    password: str | None = read_secret(
+        project_id=config_retrieve(['elasticsearch', 'password_project_id'], ''),
+        secret_name=config_retrieve(['elasticsearch', 'password_secret_id'], ''),
+        fail_gracefully=True,
+    )
+
+    # no password, but we fail gracefully
+    if password is None:
+        logging.warning(f'No permission to access ES password, skipping creation of {args.index}')
+        exit(0)
+
+    host = config_retrieve(['elasticsearch', 'host'])
+    port = config_retrieve(['elasticsearch', 'port'])
+    username = config_retrieve(['elasticsearch', 'username'])
+    logging.info(f'Connecting to ElasticSearch: host="{host}", port="{port}", user="{username}"')
+
+    ncpu = config_retrieve(['workflow', 'ncpu'], 4)
+    hl.context.init_spark(master=f'local[{ncpu}]', quiet=True)
+    hl.default_reference('GRCh38')
+
+    mt = hl.read_matrix_table(args.mt_path)
+
+    logging.info('Getting rows and exporting to the ES')
+
+    # get the rows, flattened, stripped of key and VEP annotations
+    row_ht = elasticsearch_row(mt)
+
+    es_shards = _mt_num_shards(mt)
+
+    es_client = ElasticsearchClient(host=host, port=port, es_username=username, es_password=password)
+
+    # delete the index if it exists already
+    if es_client.es.indices.exists(index=args.index):
+        es_client.es.indices.delete(index=args.index)
+
+    es_client.export_table_to_elasticsearch(row_ht, index_name=args.index, num_shards=es_shards, write_null_values=True)
+
+    if es_shards < 25:
+        es_client.wait_for_shard_transfer(args.index)
+
+    with to_path(args.flag).open('w') as f:
+        f.write('done')
+
+
+def elasticsearch_row(mt: hl.MatrixTable):
+    """
+    Prepares the mt for export.
+    - Flattens nested structs
+    - drops locus and alleles key
+    Borrowed from:
+    https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/main/luigi_pipeline/lib/model/seqr_mt_schema.py
+    """
+    # Converts a mt to the row equivalent.
+    table = mt.rows()
+    if 'vep' in table.row:
+        table = mt.drop('vep')
+    key = table.key
+    # Converts nested structs into one field, e.g. {a: {b: 1}} => a.b: 1
+    flat_table = table.flatten()
+    # When flattening, the table is unkeyed, which causes problems because our row keys should not
+    # be normal fields. We can also re-key, but I believe this is computational?
+    # PS: row key is often locus and allele, but does not have to be
+    flat_table = flat_table.drop(*key)
+    flat_table.describe()
+    return flat_table
+
+
+def _mt_num_shards(mt):
+    """
+    Calculate the number of shards from the number of variants and sequencing groups.
+    """
+    denominator = 1.4 * 10**9
```

Comment:
Any idea where this comes from? We do get ES warnings about large shards so I wonder if we should tweak?

Reply:
I'll crawl around and reference the original source for all the bits I stole, might be able to find some explanations in doctrines